npk/src/manifest.rs

```diff
@@ -12,23 +12,108 @@
 //   See the License for the specific language governing permissions and
 //   limitations under the License.
 
+use itertools::Itertools;
 use serde::{
     de::{Deserializer, Visitor},
     Deserialize, Serialize,
 };
 use serde_with::skip_serializing_none;
 use std::{
     collections::{HashMap, HashSet},
-    fmt, io,
-    path::PathBuf,
+    convert::TryFrom,
+    fmt,
+    fmt::{Display, Formatter},
+    io,
+    ops::Deref,
+    path::{Component, Component::RootDir, PathBuf},
     str::FromStr,
 };
 use thiserror::Error;
 
-pub type Name = String;
+// TODO: move out of manifest.rs
+#[derive(Clone, Eq, PartialOrd, PartialEq, Debug, Hash, Serialize, Deserialize)]
+pub struct Name(String);
+
+#[derive(Error, Clone, Eq, PartialEq, Debug, Serialize, Deserialize)]
+pub enum NameError {
+    #[error("Invalid character(s) in name")]
+    NameError,
+}
+
+impl Display for Name {
+    fn fmt(&self, f: &mut Formatter<'_>) -> Result<(), std::fmt::Error> {
+        write!(f, "{}", self.0)
+    }
+}
+
+impl Deref for Name {
+    type Target = String;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl TryFrom<String> for Name {
+    type Error = NameError;
+
+    fn try_from(value: String) -> Result<Self, Self::Error> {
+        if value
+            .chars()
+            .all(|c| matches!(c, '0'..='9' | 'A'..='Z' | 'a'..='z' | '.' | '_' | '-'))
+        {
+            Ok(Name(value))
+        } else {
+            Err(NameError::NameError)
+        }
+    }
+}
+
+// TODO: move out of manifest.rs
+#[derive(Clone, Eq, PartialOrd, PartialEq, Debug, Hash, Serialize, Deserialize)]
+pub struct NonNullString(String);
+
+#[derive(Error, Clone, Eq, PartialEq, Debug, Serialize, Deserialize)]
+pub enum NonNullStringError {
+    #[error("Invalid null byte in string")]
+    NonNullStringError,
```

Comment:
The stdlib names the same `InteriorNul` and adds the position. Check std::ffi::CString` Can we do the same? In fact we have almost the same restrictions as the `CString` but don't want to use it for discussed reasons (the null termination sucks).